pydemolib/nessiedemo/flink.py

```diff
@@ -0,0 +1,173 @@
+            if not os.path.exists(hadoop_dir):
+                tgz = self.__demo._asset_dir("{}tar.gz".format(dir_name))
+                if not os.path.exists(tgz):
+                    _Util.wget(hadoop_url, tgz)
+                _Util.exec_fail(["tar", "-x", "-z", "-C", os.path.abspath(os.path.join(hadoop_dir, "..")), "-f", tgz])
+        else:
+            raise Exception("configuration does not define hadoop.tarball. Unable to find Hadoop.")
+
+        print("Using Hadoop in {}".format(hadoop_dir))
+
+        os.environ["HADOOP_HOME"] = hadoop_dir
+        self.__copy_all_hadoop_jars_to_pyflink()
+
+    def __copy_all_hadoop_jars_to_pyflink(self: T) -> T:
+        jar_files = []
+
+        for root, _, files in os.walk(os.getenv("HADOOP_HOME")):
+            for file in files:
+                if file.endswith(".jar"):
+                    jar_files.append(os.path.join(root, file))
+
+        pyflink_lib_dir = os.path.join(site.getsitepackages()[0], "pyflink", "lib")
+        if not os.path.exists(pyflink_lib_dir):
+            raise Exception("could not find pyflink lib directory at location " + pyflink_lib_dir)
+
+        print(f"Copying all HADOOP jar files into the pyflink lib dir at location {pyflink_lib_dir}")
+        for jar in jar_files:
+            shutil.copy(jar, pyflink_lib_dir)
+
+        return self
+
+    def __enter__(self: T) -> T:
+        """Noop."""
+        return self
+
+    def __exit__(self: T, exc_type: type, exc_val: BaseException, exc_tb: TracebackType) -> None:
+        """Calls `stop()` on the `NessieDemo` instance."""
+        self.dispose()
+
+    def get_flink_warehouse(self: T) -> str:
+        """Get the path to for the 'Flink Warehouse'."""
+        return "file://{}".format(self.__demo._asset_dir("flink_warehouse"))
+
+    def __execution_env(self: T) -> StreamExecutionEnvironment:
+        if hasattr(self, "__env"):
+            return getattr(self, "__env")
+        env = StreamExecutionEnvironment.get_execution_environment()
+
+        iceberg_flink_runtime_jar = "iceberg-flink-runtime-{}.jar".format(self.__demo.get_iceberg_version())
+        asset_dir = self.__demo._asset_dir(iceberg_flink_runtime_jar)
+        _Util.wget(self.__demo.get_iceberg_download_url_for_jar("iceberg-flink-runtime"), asset_dir)
+        env.add_jars("file://{}".format(asset_dir))
+
+        setattr(self, "__env", env)
+        return env
+
+    def table_env_for_ref(self: T, nessie_ref: str, catalog_name: str = "nessie") -> StreamTableEnvironment:
+        """Retrieve a new StreamTableEnvironment ready to use against the given Nessie reference.
+
+        :param nessie_ref: the Nessie reference to configure in the StreamTableEnvironment.
+        Can be a branch name, tag name or commit hash.
+        :param catalog_name: The catalog name to use. Defaults to `nessie`
+        :return: new StreamTableEnvironment
+        """
+        table_env = StreamTableEnvironment.create(self.__execution_env())
```

Comment:
Does `StreamTableEnvironment` need to be closed? If yes, let's at least note it in the function doc.

Reply:
I checked and there's nothing that needs closing on it